### PR TITLE
Documentation: new throttling syntax, API and configuration

### PR DESCRIPTION
**Pull request by shlomi-noach**
This PR documents changes in:

- https://github.com/vitessio/vitess/pull/10198: new command line flag `vttablet --heartbeat_on_demand_duration`
- https://github.com/vitessio/vitess/pull/10162 (not merged yet, hence this PR is draft): new syntax `ALTER VITESS_MIGRATION [UN]THROTTLE ...`

It also adds documentation for:

- `/throttler` API endpoints `throttle-app`, `unthrottle-app`, `throttled-apps` -- APIs which were added but not documented a while back
- `--heartbeat_interval` and general discussion about the relationship between the throttler and heartbeats

See also: https://github.com/vitessio/vitess/issues/10196

### CODE REVIEW

**Comment by shlomi-noach** [comment]
https://github.com/vitessio/vitess/pull/10162 is merged and so this PR is now ready to review

**Comment by shlomi-noach** [comment]
Call for review

File: content/en/docs/14.0/reference/features/tablet-throttler.md [hunk 1]
```diff
@@ -99,14 +99,14 @@ Vitess only supports gauges for custom metrics: the user may define a query whic
...
-- To set the tablet types that the throttler queries for lag, use the `vttablet` flag `-throttle_tablet_types="replica,rdonly"`. The default tablet type is `replica`; this type is always implicitly included in the tablet types list. You may add any other tablet type. Any type not specified is ignored by the throttler.
-- To override the default lag evaluation, and measure a different metric, use `-throttle_metrics_query`. The query must be either of these forms:
+- Use the `vttablet` flag `--throttle_threshold` to set a lag threshold value. The default threshold is `1sec` and is set upon tablet startup. For example, to set a half-second lag threshold, use the flag `--throttle_threshold=0.5s`.
+- To set the tablet types that the throttler queries for lag, use the `vttablet` flag `--throttle_tablet_types="replica,rdonly"`. The default tablet type is `replica`; this type is always implicitly included in the tablet types list. You may add any other tablet type. Any type not specified is ignored by the throttler.
```
Comment (mattlord) [inline content/en/docs/14.0/reference/features/tablet-throttler.md:103]: Side note, I wonder if we should change the default or add a note about including `primary` tablets and using `--throttle_check_as_check_self` when using throttling with OnlineDDL? Not something we should pile onto this PR here, but noting it for our future selves after seeing users struggle a little bit to figure this out on their own.

File: content/en/docs/14.0/reference/features/tablet-throttler.md [hunk 2]
```diff
@@ -117,17 +117,25 @@ $ vttablet
...
+
+- `--heartbeat_interval` indicates how frequently heartbeats are injected. The interval should over-sample the `--throttle_threshold`. For example, if `--throttle_threshold` is `1s`, then `--heartbeat_interval` should be `250ms` or less.
+- `--heartbeat_on_demand_duration` ensures heartbeats are only injected when needed. Heartbeats are written to the binary logs, and can therefore bloat them. If this is a concern, configure for example: `--heartbeat_on_demand_duration 5s`. This setting means: any throttler request starts a `5s` lease of heartbeat writes.
+  in normal times, heartbeats are not written. Once a throttle check is requested (e.g. by a running migration), the throttler asks the tablet to start a `5s` lease of heartbeats. that first check is likely to return a non-OK code, because heartbeats were stale. However, subsequent checks will soon pick on the newly injected heartbeats. Checks made while the lease is ongoing further extend the lease. In the scenario of a running migration, we can expect heartbeats to begin as soon as the migration begins, and terminate `5s` (in our example) after the migration completes.
```
Comment (mattlord) [inline content/en/docs/14.0/reference/features/tablet-throttler.md:126]: 1. only injected when needed (e.g. during active VReplication workflows such as MoveTables or OnlineDDL).
2. I think this is a new sentence, In normal times....
3. will soon pick *up* on...
4. while the lease is held, further extend the lease time

File: content/en/docs/14.0/reference/features/tablet-throttler.md [hunk 3]
```diff
@@ -141,9 +149,25 @@ A `HEAD` request is sufficient. A `GET` request also provides a `JSON` output. F
 
 In the first two above examples we can see that the tablet is configured to throttle at `1sec`
 
-Tablet also provides `/throttler/status` endpoint. This is useful for monitoring and management purposes. 
+### Instructions
+
+- `/throttler/throttle-app?app=<name>[&duration=<duration>][&ratio=<ratio>][&p=low]`: instructs the throttler to begin throttling requests from given app.
+  - An optional `duration` value auto expires the throttling after indicated time. You may specify these units: `s` (seconds), `m` (minutes), `h` (hours) or combinations. Example values: `90s`, `30m`, `1h`, `1h30m`, etc.
+  - An optional `ratio` value indicates the throttling intensity, ranging from `0` (no throttling at all) to `1.0` (the default, full throttle).
```
Comment (mattlord) [inline content/en/docs/14.0/reference/features/tablet-throttler.md:156]: This is irrespective of the query and threshold? i.e. if I set this to 50% or 0.5, does that mean I'm throttling the work by 50% irrespective of other variables/factors, or does it mean that only 50% of the time I'm actually checking the throttling metric value? In that case I'm throttling by *at least* 50% vs exactly 50%? That make sense?

I believe it's the former, meaning that if I set the ratio to 1 then every request will be throttled and no work will be done, no matter what throttler metric and threshold I normally have in place. But I wanted to confirm. IMO it's worth a note about this here.

File: content/en/docs/14.0/reference/features/tablet-throttler.md [hunk 4]
```diff
@@ -141,9 +149,25 @@ A `HEAD` request is sufficient. A `GET` request also provides a `JSON` output. F
...
+#### Examples:
+
+- `/throttler/throttle-app?app=vreplication&ratio=0.75` rejects on average 3 out of 4 requests made by `vreplication`.
+- `/throttler/throttle-app?app=vreplication&ratio=0.25&duration=2h` rejects on average 1 out of 4 requests made by `vreplication` for the next `2` hours, after which the app is unthrottled.
+
+### Information
 
-**Example: Healthy primary tablet**
+- `/throttler/status` endpoint. This is useful for monitoring and management purposes.
+- `/throttler/throttled-apps` endpoint, listing all apps for which there's a throttling instruction
```
Comment (mattlord) [inline content/en/docs/14.0/reference/features/tablet-throttler.md:168]: Nice! ❤️

**Review by mattlord — APPROVED** [review]
This is SUPER useful! I'm excited about tablet throttling now. 🙂 We should enable it by default in 15.0 ASAP IMO.

I had a few nits and questions, but nothing blocking. Thanks again for doing this work and adding the docs for it! ❤️

**Comment by shlomi-noach** [comment]
Whoops I accidentally merged before processing your comments. Sorry! Will open a new PR to address.

File: content/en/docs/14.0/reference/features/tablet-throttler.md (hunk 1 above)
Comment (shlomi-noach) [reply content/en/docs/14.0/reference/features/tablet-throttler.md:103]: Good question. I think we should do a bit more writeup about common scenarios and setups.

File: content/en/docs/14.0/reference/features/tablet-throttler.md (hunk 2 above)
Comment (shlomi-noach) [reply content/en/docs/14.0/reference/features/tablet-throttler.md:126]: Sorry for merging prematurely. Will open a new PR per your comments.

File: content/en/docs/14.0/reference/features/tablet-throttler.md (hunk 3 above)
Comment (shlomi-noach) [reply content/en/docs/14.0/reference/features/tablet-throttler.md:156]: That is correct. In a value of `0.3`, on average `3` out or `10` throttler checks will be flatly denies, irrespective of actual metrics. the rest `7` out of `10` will respond according to actual metrics (which means they may be approved or denied)

File: content/en/docs/14.0/reference/features/tablet-throttler.md (hunk 1 above)
Comment (shlomi-noach) [reply content/en/docs/14.0/reference/features/tablet-throttler.md:103]: @mattlord again, good point. I think we should invest some time in establishing a recommended setup. Let's discuss.

**Comment by shlomi-noach** [comment]
followup in https://github.com/vitessio/website/pull/1031

**Pull request merged**